src/models/aae.jl

```diff
@@ -0,0 +1,282 @@
+"""
+function StatsBase.fit!(model::AAE, data::Tuple; max_train_time=82800, lr=0.001, 
+	batchsize=64, patience=30, check_interval::Int=10, kwargs...)
+	history = MVHistory()
+	aeopt = ADAM(lr)
+	dopt = ADAM(lr)
+	gopt = ADAM(lr)
+
+	tr_model = deepcopy(model)
+	aeps = (typeof(tr_model.prior) <: GenerativeModels.VAMP) ? 
+		Flux.params(tr_model.prior, tr_model.encoder, tr_model.decoder) : 
+		Flux.params(tr_model.encoder, tr_model.decoder)
+	dps = Flux.params(tr_model.discriminator)
+	gps = Flux.params(tr_model.encoder)
+	_patience = patience
+
+	tr_x = data[1][1]
+	val_x = data[2][1][:,data[2][2] .== 0]
+	val_N = size(val_x,2)
+
+	# on large datasets, batching loss is faster
+	best_val_loss = Inf
+	i = 1
+	start_time = time() # end the training loop after 23hrs
+	for batch in RandomBatches(tr_x, batchsize)
+		# ae loss
+		batch_aeloss = 0f0
+		gs = gradient(() -> begin 
+			batch_aeloss = aeloss(tr_model,batch)
+		end, aeps)
+	 	Flux.update!(aeopt, aeps, gs)
+
+	 	# disc loss
+		batch_dloss = 0f0
+		gs = gradient(() -> begin 
+			batch_dloss = dloss(tr_model,batch)
+		end, dps)
+	 	Flux.update!(dopt, dps, gs)
+
+	 	# gen loss
+		batch_gloss = 0f0
+		gs = gradient(() -> begin 
+			batch_gloss = gloss(tr_model,batch)
+		end, gps)
+	 	Flux.update!(gopt, gps, gs)
+
+		# validation
+		val_loss = (val_N > 5000) ? loss(tr_model, val_x, 256) : loss(tr_model, val_x)
+		(i%check_interval == 0) ? (@info "$i - loss: $(batch_aeloss) (autoencoder) $(batch_dloss) (discriminator) $(batch_gloss) (generator)  | $(val_loss) (validation)") : nothing
+		
+		# check nans
+		if isnan(val_loss) || isnan(batch_aeloss) || isnan(batch_dloss) || isnan(batch_gloss)
+			error("Encountered invalid values in loss function.")
+		end
+
+		# save training progress
+		push!(history, :training_aeloss, i, batch_aeloss)
+		push!(history, :training_dloss, i, batch_dloss)
+		push!(history, :training_gloss, i, batch_gloss)
+		push!(history, :validation_loss, i, val_loss)
+			
+		# early stopping
+		if val_loss < best_val_loss
+			best_val_loss = val_loss
+			_patience = patience
+
+			# this should save the model at least once
+			# when the validation loss is decreasing 
+			if mod(i, 10) == 0
+				model = deepcopy(tr_model)
+			end
+		elseif time() - start_time > max_train_time # stop early if time is running out
+			model = deepcopy(tr_model)
+			@info "Stopped training after $(i) iterations, $((time() - start_time)/3600) hours."
+			break
+		else # else stop if the model has not improved for `patience` iterations
+			_patience -= 1
+			if _patience == 0
+				@info "Stopped training after $(i) iterations."
+				break
+			end
+		end
+		i += 1
+	end
+	# again, this is not optimal, the model should be passed by reference and only the reference should be edited
+	(history=history, iterations=i, model=model, npars=sum(map(p->length(p), Flux.params(model))))
+end
+
+"""
+	AEModel
+
+A Union of VAE and AAE types.
+"""
+AEModel = Union{VAE, AAE}
```

Comment:
This may be not the best place for the shared code, but let's leave it here.

Reply:
Yeah it's not really a lot of code so I thought it does not really need a separate file.